# Ecommerce Dataset

In [3]:
import pandas as pd
import numpy as np
import os
import plotly.express as px

### `Load the Dataset`

In [2]:
df = pd.read_csv(os.path.join(os.getcwd(),"ecommerce.csv"))
df.head()

InvoiceNo    UnitPrice  OrderValue  Quantit            Country  \
0       549185       0.85       10.20       12  United Kingdom      
1       576381       2.95       35.40       12  United Kingdom      
2       551192       1.25       20.00       16  United Kingdom      
3       573553       7.46        7.46        1  United Kingdom      
4       539436       2.51        2.51        1  United Kingdom      

    InvoiceDate  InvoiceTime  Year-Month Major Category  Minor Category  \
0     7/4/11        09:35     2011-04         Clothes              Tops   
1   15/11/11        09:26     2011-11         Clothes             Shoes   
2    27/4/11        10:54     2011-04         Kitchen           Cutlery   
3   31/10/11        13:48     2011-10          Garden              Turf   
4   17/12/10        14:49     2010-12          Garden             Hoses   

                           Description  
0     PACK OF 20 NAPKINS PANTRY DESIGN  
1      NATURAL SLATE HEART CHALKBOARD   
2               36 PENCILS TUBE SKULLS  
3   SET 6 SCHOOL MILK BOTTLES IN CRATE  
4                   FINE WICKER HEART

In [3]:
## Delete all spaces before and after the column name 
df.columns = df.columns.str.strip()
df.tail()

InvoiceNo  UnitPrice  OrderValue  Quantit            Country  \
695     564087       1.25        20.0       16       Hong Kong      
696     553148       0.55        13.2       24       Hong Kong      
697     564087       4.95        29.7        6       Hong Kong      
698     553148       0.85        27.2       32       Hong Kong      
699     564087       1.25        20.0       16       Hong Kong      

      InvoiceDate  InvoiceTime  Year-Month Major Category  Minor Category  \
695    23/8/11        09:38     2011-08         Clothes              Tops   
696    13/5/11        14:09     2011-05          Garden             Hoses   
697    23/8/11        09:38     2011-08          Garden             Rakes   
698    13/5/11        14:09     2011-05         Kitchen    Cooking Knives   
699    23/8/11        09:38     2011-08         Clothes              Tops   

                             Description  
695             WOOD STAMP SET THANK YOU  
696                  GREEN POLKADOT BOWL  
697                  TOY TIDY SPACEBOY    
698     RED RETROSPOT CHILDRENS UMBRELLA  
699              PACK OF 60 SPACEBOY CAK

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 700 entries, 0 to 699
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   InvoiceNo       700 non-null    int64  
 1   UnitPrice       700 non-null    float64
 2   OrderValue      700 non-null    float64
 3   Quantit         700 non-null    int64  
 4   Country         700 non-null    object 
 5   InvoiceDate     700 non-null    object 
 6   InvoiceTime     700 non-null    object 
 7   Year-Month      700 non-null    object 
 8   Major Category  700 non-null    object 
 9   Minor Category  700 non-null    object 
 10  Description     700 non-null    object 
dtypes: float64(2), int64(2), object(7)
memory usage: 60.3+ KB


In [5]:
## Year-Month,InvoiceDate must be converted to DateTime datetype 
df["Year-Month"] = pd.to_datetime(df["Year-Month"])
df['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'])

In [6]:
print(f"Number of NULL values : {df.isna().any().sum()}")
print(f"Number of duplicated values : {df.duplicated().sum()}")

Number of NULL values : 0
Number of duplicated values : 0


##### `There are no missing values and duplicated values`

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 700 entries, 0 to 699
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   InvoiceNo       700 non-null    int64         
 1   UnitPrice       700 non-null    float64       
 2   OrderValue      700 non-null    float64       
 3   Quantit         700 non-null    int64         
 4   Country         700 non-null    object        
 5   InvoiceDate     700 non-null    datetime64[ns]
 6   InvoiceTime     700 non-null    object        
 7   Year-Month      700 non-null    datetime64[ns]
 8   Major Category  700 non-null    object        
 9   Minor Category  700 non-null    object        
 10  Description     700 non-null    object        
dtypes: datetime64[ns](2), float64(2), int64(2), object(5)
memory usage: 60.3+ KB


### 1- What are the total Order value for each country ?

In [8]:
df.groupby("Country")[["OrderValue"]].sum().sort_values(by="OrderValue", ascending=False).reset_index()

Country  OrderValue
0       Australia        6349.89
1          France        5638.16
2  United Kingdom        5582.93
3         Germany        3153.71
4       Hong Kong        1660.24

### 2- What each Major Category sold quantity during November 2011?

In [9]:
df_november_2011 = df[df["Year-Month"]=="2011-11"]
df_november_2011.groupby("Major Category")[["Quantit"]].sum().sort_values(by="Quantit",ascending=True).reset_index()

Major Category  Quantit
0    Household        246
1      Clothes        280
2       Garden        449
3      Kitchen        538

### 3- What's Total Order Value for each Minor Item in Kitchen?


In [10]:
df_kitchen = df.groupby(["Major Category","Minor Category"])[["OrderValue"]].sum() \
                                                                            .reset_index() \
                                                                            .loc[7:12,["Minor Category","OrderValue"]]
df_kitchen.sort_values(by="OrderValue").reset_index(drop=True)

Minor Category  OrderValue
0        Coasters      324.30
1         Cutlery      960.29
2  Cooking Knives     1055.46
3          Plates     1148.67
4           Bowls     1227.39
5          Scales     1691.79

##### `Note` : The value of orders minimun in the products "Coasters"  , The value of orders maximum in the products "Scales"

### 4- What Month/s contributes to high demand in orders ?


In [11]:
df["Year-Month"] = df["Year-Month"].dt.strftime('%Y-%m')
df.groupby("Year-Month")[["Quantit"]].sum().sort_values(by="Quantit",ascending=False).reset_index()

Year-Month  Quantit
0     2011-09     2002
1     2011-06     1733
2     2011-05     1668
3     2011-03     1660
4     2011-11     1513
5     2011-08     1251
6     2011-01     1184
7     2011-02     1132
8     2011-07      987
9     2011-10      693
10    2010-12      397
11    2011-12      201
12    2011-04      185

##### in `September` month is high demand in orders .

### 5- What Month/s contributes to high OrderValue?


In [12]:
df.groupby(("Year-Month"))[["OrderValue"]].sum().sort_values(by="OrderValue",ascending=False).reset_index()

Year-Month  OrderValue
0     2011-11     2553.74
1     2011-05     2394.34
2     2011-09     2339.75
3     2011-08     2327.19
4     2011-03     2324.19
5     2011-01     2080.25
6     2011-06     1878.89
7     2011-02     1758.76
8     2011-07     1562.23
9     2011-10     1382.54
10    2010-12      795.34
11    2011-04      588.83
12    2011-12      398.88

##### `In November 2011, with a value of 2553.74`

### 6- Why April and December suffer from low sales?

In [13]:
df["month_name"] = pd.to_datetime(df["Year-Month"]).dt.month_name()
df_april_dec = df[(df["month_name"] == "April") | (df["month_name"] == "December")] 
df_april_dec.groupby(["month_name","Quantit"])[["OrderValue"]].sum()\
                                                            .reset_index()

month_name  Quantit  OrderValue
0       April        1      240.51
1       April        2       19.20
2       April        4       59.20
3       April        6       22.80
4       April        8       29.12
5       April       10       49.80
6       April       12      113.40
7       April       16       20.00
8       April       24       34.80
9    December        1      122.20
10   December        2       62.54
11   December        3       70.23
12   December        4      133.28
13   December        6      137.40
14   December        7       44.03
15   December        8       70.40
16   December       10       20.80
17   December       12      134.52
18   December       20       33.40
19   December       24      136.32
20   December       50      147.50
21   December       96       81.60

In [14]:
df_april_dec.groupby(["Year-Month","Major Category","Minor Category","Country"])[["OrderValue"]].sum().reset_index()

Year-Month Major Category  Minor Category            Country  OrderValue
0     2010-12       Garden             Hoses  United Kingdom           6.75
1     2010-12       Garden             Rakes          France          10.08
2     2010-12       Garden             Rakes  United Kingdom          30.90
3     2010-12       Garden             Seeds  United Kingdom           4.95
4     2010-12      Clothes              Hats  United Kingdom          35.12
..        ...            ...             ...                ...         ...
61    2011-12    Household              Rugs         Germany          55.08
62    2011-12    Household             Lamps  United Kingdom           8.97
63    2011-12    Household            Chairs          France          16.60
64    2011-12    Household            Chairs  United Kingdom          20.80
65    2011-12    Household          Curtains  United Kingdom          21.00

[66 rows x 5 columns]

### 7- Why May and November have high sales? 


In [15]:
df_nov_may = df[(df["month_name"] == "May") | (df["month_name"] == "November")] 
df_nov_may.groupby(["month_name","Quantit"])[["OrderValue"]].sum()\
                                                            .reset_index()

month_name  Quantit  OrderValue
0         May        1       11.37
1         May        2       89.94
2         May        3       17.85
3         May        4       96.44
4         May        6      224.40
5         May        8      175.60
6         May       10      221.50
7         May       12      134.04
8         May       14       69.44
9         May       20       42.00
10        May       24      115.68
11        May       30       63.00
12        May       32       54.40
13        May       36      178.20
14        May       48      110.40
15        May       60      126.00
16        May       96       36.48
17        May      144      237.60
18        May      600      390.00
19   November        1       78.07
20   November        2       67.18
21   November        3       45.33
22   November        4      222.40
23   November        5        1.45
24   November        6      183.60
25   November        7        5.95
26   November        8      133.20
27   November        9       74.16
28   November       10       95.40
29   November       12      333.36
30   November       18       59.94
31   November       20       57.80
32   November       22       20.68
33   November       24      223.68
34   November       25       10.50
35   November       30       55.80
36   November       32       81.60
37   November       60       51.00
38   November       72      128.88
39   November       96      408.96
40   November      120      214.80

In [16]:
df_nov_may.groupby(["Year-Month","Major Category","Minor Category","Country"])[["OrderValue"]].sum().reset_index()

Year-Month Major Category  Minor Category            Country  OrderValue
0     2011-05       Garden              Turf          France          10.08
1     2011-05       Garden              Turf         Germany         144.00
2     2011-05       Garden              Turf  United Kingdom          22.50
3     2011-05       Garden             Hoses          France          68.04
4     2011-05       Garden             Hoses       Hong Kong          26.40
..        ...            ...             ...                ...         ...
81    2011-11    Household            Chairs          France           8.25
82    2011-11    Household            Chairs         Germany          37.44
83    2011-11    Household            Chairs  United Kingdom          64.63
84    2011-11    Household          Curtains          France          19.96
85    2011-11    Household          Curtains  United Kingdom         111.17

[86 rows x 5 columns]

#### `Note`
###### `Quantity` in April and December is low and in May and November is high

### 8- Most Major Items sold in April and December ?


In [17]:
df_april_dec.groupby("Major Category")[["Minor Category"]].count().rename(columns={"Minor Category":"Count"})

Count
Major Category       
     Garden        26
    Clothes        22
    Kitchen        25
  Household        26

In [18]:
df_april_dec.groupby("Major Category")[["OrderValue"]].sum()

OrderValue
Major Category            
     Garden         215.86
    Clothes         347.70
    Kitchen         553.63
  Household         665.86

#### `Note` 
- With the best selling category in the month of April and December is garden and Household but the ordervalue in kitchen , and household is high So we will say `Household` is Most Major Items sold .

In [19]:
df_april_dec.groupby("Major Category")[["InvoiceNo","UnitPrice","OrderValue","Quantit"]].sum()

InvoiceNo  UnitPrice  OrderValue  Quantit
Major Category                                           
     Garden      14568399      86.34      215.86      142
    Clothes      12186199      77.19      347.70      160
    Kitchen      13701708      97.74      553.63      209
  Household      14420547     300.84      665.86      272

### 9- Most Major Items sold in May and November ?


In [20]:
df_nov_may.groupby("Major Category")[["Minor Category"]].count().rename(columns={"Minor Category":"Count"})\
                                                                                .sort_values(by="Count",ascending=False)

Count
Major Category       
    Kitchen        48
     Garden        47
  Household        41
    Clothes        36

In [21]:
df_nov_may.groupby("Major Category")[["OrderValue"]].sum().sort_values(by="OrderValue",ascending=False)

OrderValue
Major Category            
     Garden        1651.90
    Kitchen        1523.03
    Clothes         936.62
  Household         836.53

##### `Note` most major Items sold in May and November is Garden and Kitchen

In [22]:
df_nov_may.groupby("Major Category")[["InvoiceNo","UnitPrice","OrderValue","Quantit"]].sum()

InvoiceNo  UnitPrice  OrderValue  Quantit
Major Category                                           
     Garden      26754139     151.19     1651.90     1234
    Clothes      20393767     128.03      936.62      538
    Kitchen      27295187     133.72     1523.03      876
  Household      23229742     100.16      836.53      533

### 10- What Hour recives high demand for ordering?

In [23]:
h = []
for i in df['InvoiceTime']:
    h.append(int(i.split(":")[0]))

In [24]:
df["Hours"] = pd.DataFrame(h) 

In [25]:
df.groupby("Hours")[["Quantit"]].sum().sort_values(by="Quantit", ascending=False).reset_index()

Hours  Quantit
0      13     2958
1      10     2663
2      14     2016
3      11     1661
4      15     1448
5       9     1398
6      12     1288
7       8      624
8      16      331
9      17      116
10      7       96
11     18        4
12     19        2
13     20        1

#####  `13 PM , recives high demand for ordering `

### 11- What's the Time of the day that contributes to high Order Values?

In [26]:
ordertime = []
for i in df["Hours"]:
    if (i > 5) and (i < 12)  : 
        ordertime.append("Morning")
    elif (i >= 12) and (i < 16)  : 
        ordertime.append("Afternoon")
    else :
        ordertime.append("Evening")

In [27]:
df["OrderTimeOfDay"] = pd.DataFrame(ordertime)

In [28]:
df.groupby("OrderTimeOfDay")[["Quantit"]].sum().sort_values(by="Quantit", ascending=False)

Quantit
OrderTimeOfDay         
Afternoon          7710
Morning            6442
Evening             454

##### the Time of the day that contributes to high Order Values is `Afternoon`. 


### 12 - Order value For each Country in each month

In [29]:
round(pd.pivot_table(data=df, values="OrderValue", index="Year-Month", columns="Country", aggfunc="sum"),2)

Country             France            Germany          Australia     \
Year-Month                                                            
2010-12                128.33             227.70             147.50   
2011-01                576.29             279.10             471.30   
2011-02                213.60             420.50             511.94   
2011-03                508.47              50.55            1360.70   
2011-04                 30.30              98.00                NaN   
2011-05                392.82             753.30             589.38   
2011-06                514.00             138.38             748.00   
2011-07                244.86             221.99             477.26   
2011-08                839.34             137.55             558.18   
2011-09                379.58             217.26            1007.97   
2011-10                577.14             167.46             231.90   
2011-11               1108.31             356.84             245.76   
2011-12                125.12              85.08                NaN   

Country          Hong Kong     United Kingdom     
Year-Month                                        
2010-12                   NaN             291.81  
2011-01                595.80             157.76  
2011-02                   NaN             612.72  
2011-03                   NaN             404.47  
2011-04                 73.10             387.43  
2011-05                269.85             388.99  
2011-06                 15.00             463.51  
2011-07                   NaN             618.12  
2011-08                578.90             213.22  
2011-09                   NaN             734.94  
2011-10                127.59             278.45  
2011-11                   NaN             842.83  
2011-12                   NaN             188.68

### 13 - Order Value For each Country and Major Category in Minor Category

In [30]:
round(pd.pivot_table(data=df, values="OrderValue", index=["Country","Major Category"], columns="Year-Month", aggfunc="sum"),2)

Year-Month                        2010-12  2011-01  2011-02  2011-03  2011-04  \
Country           Major Category                                                
        France         Garden       10.08    85.35    57.30   138.72      NaN   
                      Clothes       41.70   137.85   108.30   162.15    15.00   
                      Kitchen       18.35   197.80      NaN   192.60    15.30   
                    Household       58.20   155.29    48.00    15.00      NaN   
       Germany         Garden         NaN   127.80    23.60    35.55    20.80   
                      Clothes         NaN    97.30    15.90      NaN      NaN   
                      Kitchen      125.70    20.40      NaN    15.00     7.80   
                    Household      102.00    33.60   381.00      NaN    69.40   
     Australia         Garden         NaN    16.50    53.10    51.84      NaN   
                      Clothes         NaN      NaN    68.00   735.36      NaN   
                      Kitchen      147.50   139.20    51.84   557.60      NaN   
                    Household         NaN   315.60   339.00    15.90      NaN   
     Hong Kong         Garden         NaN    78.90      NaN      NaN    12.50   
                      Clothes         NaN   450.90      NaN      NaN    15.60   
                      Kitchen         NaN      NaN      NaN      NaN      NaN   
                    Household         NaN    66.00      NaN      NaN    45.00   
United Kingdom         Garden       42.60    49.87    68.73    53.07    48.28   
                      Clothes       60.97    66.95     9.00   272.90    80.45   
                      Kitchen      140.05     9.87    56.99    33.95    38.08   
                    Household       48.19    31.07   478.00    44.55   220.62   

Year-Month                        2011-05  2011-06  2011-07  2011-08  2011-09  \
Country           Major Category                                                
        France         Garden      168.12   303.60    24.36   667.38    87.28   
                      Clothes       48.50    65.10    30.00    92.06    53.04   
                      Kitchen      133.76    67.90    82.00    39.60   155.42   
                    Household       42.44    77.40   108.50    40.30    83.84   
       Germany         Garden      534.00    48.30    49.55    49.65    17.70   
                      Clothes       85.50    29.88   110.46    39.90   177.30   
                      Kitchen      133.80     8.50    61.98      NaN    22.26   
                    Household         NaN    51.70      NaN    48.00      NaN   
     Australia         Garden         NaN   222.00      NaN      NaN   689.10   
                      Clothes      237.60      NaN    30.60      NaN    90.27   
                      Kitchen      245.70   256.00   412.36   433.38   228.60   
                    Household      106.08   270.00    34.30   124.80      NaN   
     Hong Kong         Garden       91.15    15.00      NaN   248.10      NaN   
                      Clothes       33.50      NaN      NaN    79.60      NaN   
                      Kitchen       62.40      NaN      NaN   165.60      NaN   
                    Household       82.80      NaN      NaN    85.60      NaN   
United Kingdom         Garden       22.50    98.66   134.65    30.00   503.05   
                      Clothes      127.44   149.73    64.89    39.69    68.58   
                      Kitchen       66.58   185.03   355.26   106.23    77.22   
                    Household      172.47    30.09    63.32    37.30    86.09   

Year-Month                        2011-10  2011-11  2011-12  
Country           Major Category                             
        France         Garden       27.50   292.39    26.32  
                      Clothes      297.16   196.53    62.40  
                      Kitchen      171.60   442.98    19.80  
                    Household       80.88   176.41    16.60  
       Germany         Garden       48.00   131.90    30.00  


### 14 - Order Value For each Country and Major Category in Minor Category


In [31]:
round(pd.pivot_table(data=df, values="OrderValue", index=["Country","Major Category"], columns="Hours"),2)

Hours                                7       8       9       10      11  \
Country           Major Category                                          
        France         Garden       NaN   10.08   24.54   22.85   83.93   
                      Clothes     244.8   34.00   26.59   14.61   41.83   
                      Kitchen       NaN   28.20   29.37   22.84   55.48   
                    Household       NaN     NaN   35.40   24.43   32.88   
       Germany         Garden       NaN   34.27   45.41   19.79   16.37   
                      Clothes       NaN     NaN   32.10   14.97   13.42   
                      Kitchen       NaN     NaN   11.85   23.04   89.70   
                    Household       NaN    4.68   48.75  104.02    7.59   
     Australia         Garden       NaN     NaN   18.65   51.84     NaN   
                      Clothes       NaN     NaN   68.00  201.12   52.02   
                      Kitchen       NaN  208.44   59.67  227.13  139.20   
                    Household       NaN  124.80  176.80   16.60  157.80   
     Hong Kong         Garden       NaN     NaN   43.43   15.00     NaN   
                      Clothes       NaN     NaN   23.80     NaN     NaN   
                      Kitchen       NaN   23.40   82.80     NaN     NaN   
                    Household       NaN     NaN   32.65     NaN     NaN   
United Kingdom         Garden       NaN   36.00   16.59   16.29   11.63   
                      Clothes       NaN   16.50   16.18   10.76   20.56   
                      Kitchen       NaN   11.90   18.65   16.07   36.45   
                    Household       NaN    7.44   26.74  103.27    6.04   

Hours                                12      13      14      15     16     17  \
Country           Major Category                                                
        France         Garden     18.32   18.16   16.06   27.32   9.12  21.25   
                      Clothes     40.20   21.82   17.24   19.98  28.20    NaN   
                      Kitchen     15.34   40.30   33.20   25.34   6.41    NaN   
                    Household     17.29   15.43   24.04   18.91  18.20    NaN   
       Germany         Garden     14.10   32.70  202.38   17.58    NaN    NaN   
                      Clothes     18.46   33.95   11.85   27.56  16.50    NaN   
                      Kitchen     13.68   16.85   32.87   31.08    NaN  15.00   
                    Household     17.70   32.72   17.56   17.70    NaN  15.00   
     Australia         Garden     17.07  427.80   39.60   53.10    NaN    NaN   
                      Clothes     14.85   30.60     NaN  130.50    NaN    NaN   
                      Kitchen     19.80  128.10   82.00  288.78    NaN    NaN   
                    Household     93.60   95.30   27.00   53.04    NaN    NaN   
     Hong Kong         Garden     15.00     NaN   24.29     NaN    NaN    NaN   
                      Clothes     10.89     NaN   60.55     NaN    NaN    NaN   
                      Kitchen     14.85     NaN   15.60     NaN    NaN    NaN   
                    Household     17.52     NaN   24.80     NaN    NaN    NaN   
United Kingdom         Garden     11.09   16.08   53.83   27.01  19.24   5.41   
                      Clothes     19.87    8.33   12.08   18.85   1.78  16.33   
                      Kitchen     12.04   23.82   14.32   37.34   6.85  14.23   
                    Household     13.69   13.54   10.62    9.00   4.54   5.49   

Hours                               18    19    20  
Country           Major Category                    
        France         Garden      NaN   NaN   NaN  
                      Clothes      NaN   NaN   NaN  
                      Kitchen      NaN   NaN   NaN  
                    Household      NaN   NaN   NaN  
       Germany         Garden      NaN   NaN   NaN  
                      Clothes      NaN   NaN   NaN  
                      Kitchen      NaN   NaN   NaN  
                    Household      NaN   NaN   NaN  
     Australia         Garden      N

In [8]:
# 15- Correlation between 'Quantit','OrderValue','UnitPrice'.
px.imshow(round(df[['Quantit','OrderValue','UnitPrice']].corr(),3), text_auto=True, color_continuous_scale=px.colors.sequential.amp)

### Done